File: go/pkg/antfly/cmd/cmd/cli/table.go
/*
Copyright 2026 The Antfly Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

	http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
package cli

import (
	"fmt"
	"os"
	"path/filepath"
	"time"

	json "github.com/antflydb/antfly/go/pkg/libaf/json"
	antfly "github.com/antflydb/antfly/go/pkg/sdk"
	"github.com/spf13/cobra"
)

func addTableCommands(parent *cobra.Command) {
	tableCmd := &cobra.Command{
		Use:   "table",
		Short: "Manage tables in the database",
		Long: `The table command provides subcommands for managing tables in the database,
including creating tables, backing them up, and restoring from backups.

When called with --table/-t and no subcommand, shows details for the specified table.
Without --table, lists all tables.`,
		RunE: func(cmd *cobra.Command, args []string) error {
			tableName, _ := cmd.Flags().GetString("table")
			formatStr, _ := cmd.Flags().GetString("output")
			format, err := parseOutputFormat(formatStr)
			if err != nil {
				return err
			}

			if tableName == "" {
				tables, err := antflyClient.ListTables(cmd.Context())
				if err != nil {
					return err
				}
				if format == outputJSON {
					return writeJSON(tables)
				}
				printTableList(tables)
				return nil
			}

			table, err := antflyClient.GetTable(cmd.Context(), tableName)
			if err != nil {
				return err
			}
			if format == outputJSON {
				return writeJSON(table)
			}
			printTableStatus(table)
			return nil
		},
	}
	tableCmd.PersistentFlags().StringP("table", "t", "", "Name of the table")
	tableCmd.Flags().StringP("output", "o", "table", "Output format: table, json")

	tableCreateCmd := &cobra.Command{
		Use:   "create",
		Short: "Creates a new table",
		Long: `Creates a new table with the specified name and number of shards.

Example with indexes:
  antfly table create --table products --index '{"name":"product_embeddings","field":"description","embedder":{"provider":"antfly","model":"BAAI/bge-small-en-v1.5"}}'

Example with config file:
  antfly table create --table products --file table-config.json

The config file should be a JSON object with optional "indexes", "schema", and "shards" fields:
  {
    "indexes": {
      "title_body": {"type": "embeddings", "template": "{{title}} {{body}}", ...}
    },
    "shards": 1,
    "schema": {}
  }`,
		RunE: func(cmd *cobra.Command, args []string) error {
			tableName, _ := cmd.Flags().GetString("table")
			if tableName == "" {
				return fmt.Errorf("--table is required")
			}
			numShards, _ := cmd.Flags().GetInt("shards")
			schemaJSON, _ := cmd.Flags().GetString("schema")
			indexStrings, _ := cmd.Flags().GetStringArray("index")
			configFile, _ := cmd.Flags().GetString("file")

			var indexes map[string]antfly.IndexConfig

			if configFile != "" {
				if len(indexStrings) > 0 {
					return fmt.Errorf("cannot specify both --file and --index")
				}
				data, err := os.ReadFile(filepath.Clean(configFile))
				if err != nil {
					return fmt.Errorf("reading config file %s: %w", configFile, err)
				}
				var fileConfig struct {
					Indexes map[string]antfly.IndexConfig `json:"indexes"`
					Schema  json.RawMessage               `json:"schema"`
					Shards  int                           `json:"shards"`
				}
				if err := json.Unmarshal(data, &fileConfig); err != nil {
					return fmt.Errorf("parsing config file %s: %w", configFile, err)
				}
				indexes = fileConfig.Indexes
				if fileConfig.Shards > 0 {
					numShards = fileConfig.Shards
				}
				if len(fileConfig.Schema) > 0 && schemaJSON == "" {
					schemaJSON = string(fileConfig.Schema)
				}
			} else if len(indexStrings) > 0 {
				indexes = make(map[string]antfly.IndexConfig, len(indexStrings))
				for _, indexStr := range indexStrings {
					var indexConf antfly.IndexConfig
					if err := json.UnmarshalString(indexStr, &indexConf); err != nil {
						return fmt.Errorf("invalid JSON for index: %w", err)
					}
					name := indexConf.Name
					if name == "" {
						return fmt.Errorf("index definition must have a 'name' field of type string")
					}
					indexes[name] = indexConf
				}
			}

			err := antflyClient.CreateTable(cmd.Context(), tableName, numShards, schemaJSON, indexes)
			if err != nil {
				return fmt.Errorf("create table failed: %w", err)
			}
			fmt.Fprintln(os.Stderr, "Create table command successful.")
			return nil
		},
	}

	tableDropCmd := &cobra.Command{
		Use:   "drop",
		Short: "Drops an existing table",
		Long:  `Drops an existing table and all its data. This operation cannot be undone.`,
		RunE: func(cmd *cobra.Command, args []string) error {
			tableName, _ := cmd.Flags().GetString("table")
			if tableName == "" {
				return fmt.Errorf("--table is required")
			}

			err := antflyClient.DropTable(cmd.Context(), tableName)
			if err != nil {
				return fmt.Errorf("drop table failed: %w", err)
			}
			fmt.Fprintln(os.Stderr, "Drop table command successful.")
			return nil
		},
	}

	tableListCmd := &cobra.Command{
		Use:   "list",
		Short: "Lists all tables",
		Long:  `Lists all tables in the database with their basic information.`,
		RunE: func(cmd *cobra.Command, args []string) error {
			formatStr, _ := cmd.Flags().GetString("output")
			format, err := parseOutputFormat(formatStr)
			if err != nil {
				return err
			}

			tables, err := antflyClient.ListTables(cmd.Context())
			if err != nil {
				return fmt.Errorf("list tables failed: %w", err)
			}
			if format == outputJSON {
				return writeJSON(tables)
			}
			printTableList(tables)
			return nil
		},
	}
	tableListCmd.Flags().StringP("output", "o", "table", "Output format: table, json")

	tableCmd.AddCommand(tableCreateCmd)
	tableCmd.AddCommand(tableDropCmd)
	tableCmd.AddCommand(tableListCmd)

	// create-table flags
	tableCreateCmd.Flags().Int("shards", 0, "Number of shards for the table")
	tableCreateCmd.Flags().String("schema", "", "JSON schema definition")
	tableCreateCmd.Flags().StringArray("index", []string{}, "JSON index definition (can be specified multiple times)")
	tableCreateCmd.Flags().StringP("file", "f", "", "Path to a JSON config file with indexes, schema, and shards")

	backupCmd := newBackupCmd()
	restoreCmd := newRestoreCmd()

	parent.AddCommand(tableCmd)
	parent.AddCommand(backupCmd)
	parent.AddCommand(restoreCmd)
}

func newBackupCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "backup",
		Short: "Backs up tables to a storage location",
		Long: `Backs up one or more tables to a given location with a unique backup ID.

Single table backup:
  antfly backup --table users --backup-id my-backup --location file:///tmp/backups

Bulk backup (all tables):
  antfly backup --backup-id cluster-backup --location file:///tmp/backups

Bulk backup (selected tables):
  antfly backup --tables users,products --backup-id selective-backup --location s3://bucket/path

List available backups:
  antfly backup --list --location file:///tmp/backups`,
		PreRunE: func(cmd *cobra.Command, args []string) error {
			return initClient(cmd, longTimeoutHTTPClient())
		},
		RunE: func(cmd *cobra.Command, args []string) error {
			listBackups, _ := cmd.Flags().GetBool("list")
			location, _ := cmd.Flags().GetString("location")

			// Handle --list flag
			if listBackups {
				if err := antflyClient.ListBackups(cmd.Context(), location); err != nil {
					return fmt.Errorf("list backups failed: %w", err)
				}
				return nil
			}

			tableName, _ := cmd.Flags().GetString("table")
			tablesStr, _ := cmd.Flags().GetString("tables")
			backupID, _ := cmd.Flags().GetString("backup-id")

			// Single table backup (--table flag provided)
			if tableName != "" {
				if backupID == "" {
					backupID = fmt.Sprintf("backup-%s-%s", tableName, time.Now().Format("20060102150405"))
				}
				if err := antflyClient.Backup(cmd.Context(), tableName, backupID, location); err != nil {
					return fmt.Errorf("backup failed: %w", err)
				}
				fmt.Fprintln(os.Stderr, "Backup command successful.")
				return nil
			}

			// Bulk backup (no --table flag)
			if backupID == "" {
				backupID = fmt.Sprintf("cluster-backup-%s", time.Now().Format("20060102150405"))
			}

			if err := antflyClient.ClusterBackup(cmd.Context(), backupID, location, splitCSV(tablesStr)); err != nil {
				return fmt.Errorf("backup failed: %w", err)
			}
			fmt.Fprintln(os.Stderr, "Backup command successful.")
			return nil
		},
	}

	cmd.Flags().StringP("table", "t", "", "Name of a single table to backup")
	cmd.Flags().String("tables", "", "Comma-separated list of tables to backup (bulk mode)")
	cmd.Flags().String("backup-id", "", "Unique ID for this backup")
	cmd.Flags().String("location", "file:///tmp/antfly_backups", "Backup location (e.g., file:///path/to/dir or s3://bucket/path)")
	cmd.Flags().Bool("list", false, "List available backups at the location")

	return cmd
}

func newRestoreCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "restore",
		Short: "Restores tables from a backup",
		Long: `Restores one or more tables from a backup.

Single table restore:
  antfly restore --table users --backup-id my-backup --location file:///tmp/backups

Bulk restore (all tables from backup):
  antfly restore --backup-id cluster-backup --location file:///tmp/backups

Bulk restore (selected tables):
  antfly restore --tables users,products --backup-id cluster-backup --location s3://bucket/path

Restore modes (for bulk restore):
  - fail_if_exists: Abort if any table already exists (default)
  - skip_if_exists: Skip existing tables, restore others
  - overwrite: Drop and recreate existing tables`,
		PreRunE: func(cmd *cobra.Command, args []string) error {
			return initClient(cmd, longTimeoutHTTPClient())
		},
		RunE: func(cmd *cobra.Command, args []string) error {
			tableName, _ := cmd.Flags().GetString("table")
			tablesStr, _ := cmd.Flags().GetString("tables")
			backupID, _ := cmd.Flags().GetString("backup-id")
			location, _ := cmd.Flags().GetString("location")
			restoreMode, _ := cmd.Flags().GetString("mode")

			if backupID == "" {
				return fmt.Errorf("--backup-id is required for restore command")
			}

			// Single table restore (--table flag provided)
			if tableName != "" {
				if err := antflyClient.Restore(cmd.Context(), tableName, backupID, location); err != nil {
					return fmt.Errorf("restore failed: %w", err)
				}
				fmt.Fprintln(os.Stderr, "Restore command successfully initiated. It may take some time for the table to become fully available.")
				return nil
			}

			// Bulk restore (no --table flag)
			if err := antflyClient.ClusterRestore(cmd.Context(), backupID, location, splitCSV(tablesStr), restoreMode); err != nil {
				return fmt.Errorf("restore failed: %w", err)
			}
			fmt.Fprintln(os.Stderr, "Restore command successfully initiated. It may take some time for the tables to become fully available.")
			return nil
		},
	}

	cmd.Flags().StringP("table", "t", "", "Name of a single table to restore")
	cmd.Flags().String("tables", "", "Comma-separated list of tables to restore (bulk mode)")
	cmd.Flags().String("backup-id", "", "ID of the backup to restore from (required)")
	cmd.Flags().String("location", "file:///tmp/antfly_backups", "Location of the backup (e.g., file:///path/to/dir or s3://bucket/path)")
	cmd.Flags().String("mode", "fail_if_exists", "Restore mode for bulk restore: fail_if_exists, skip_if_exists, overwrite")

	return cmd
}

File: zig/pkg/antfly/src/cmd/cli/table.zig
// Copyright 2026 Antfly, Inc.
//
// Licensed under the Elastic License 2.0 (ELv2); you may not use this file
// except in compliance with the Elastic License 2.0. You may obtain a copy of
// the Elastic License 2.0 at
//
//     https://www.antfly.io/licensing/ELv2-license
//
// Unless required by applicable law or agreed to in writing, software distributed
// under the Elastic License 2.0 is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// Elastic License 2.0 for the specific language governing permissions and
// limitations.

const std = @import("std");
const ant_json = @import("antfly-json");
const antfly_client = @import("antfly-client");
const httpx = antfly_client.httpx;
const cli = @import("mod.zig");

const TableCreateOption = enum { table, shards, file, index, schema };

fn tableCreateOption(arg: []const u8) ?TableCreateOption {
    if (std.mem.eql(u8, arg, "--table") or std.mem.eql(u8, arg, "-t")) return .table;
    if (std.mem.eql(u8, arg, "--shards")) return .shards;
    if (std.mem.eql(u8, arg, "--file") or std.mem.eql(u8, arg, "-f")) return .file;
    if (std.mem.eql(u8, arg, "--index")) return .index;
    if (std.mem.eql(u8, arg, "--schema")) return .schema;
    return null;
}

fn parseInlineSchemaConfig(allocator: std.mem.Allocator, raw: []const u8) !antfly_client.types.TableSchema {
    return std.json.parseFromSliceLeaky(
        antfly_client.types.TableSchema,
        allocator,
        raw,
        .{ .allocate = .alloc_always },
    );
}

const InlineIndexConfig = struct {
    name: []const u8,
    request: antfly_client.types.CreateIndexRequest,
};

fn parseInlineIndexConfig(allocator: std.mem.Allocator, raw: []const u8) !InlineIndexConfig {
    var value = try std.json.parseFromSliceLeaky(std.json.Value, allocator, raw, .{
        .allocate = .alloc_always,
    });
    const object = switch (value) {
        .object => |*object| object,
        else => return error.UnexpectedToken,
    };
    const name_value = object.get("name") orelse return error.MissingField;
    const name = switch (name_value) {
        .string => |string| string,
        else => return error.UnexpectedToken,
    };
    _ = object.swapRemove("name");
    return .{
        .name = name,
        .request = try std.json.parseFromValueLeaky(
            antfly_client.types.CreateIndexRequest,
            allocator,
            value,
            .{},
        ),
    };
}

fn putInlineIndexConfig(
    allocator: std.mem.Allocator,
    indexes: *std.json.ArrayHashMap(antfly_client.types.CreateIndexRequest),
    index: InlineIndexConfig,
) !void {
    if (index.name.len == 0) return error.EmptyIndexName;
    if (indexes.map.contains(index.name)) return error.DuplicateIndexName;
    try indexes.map.put(allocator, index.name, index.request);
}

pub fn run(allocator: std.mem.Allocator, io: std.Io, client: *antfly_client.AntflyClient, args: *std.process.Args.Iterator) !void {
    const subcommand = args.next() orelse {
        return listTables(allocator, io, client);
    };

    if (std.mem.eql(u8, subcommand, "create")) return createTable(allocator, io, client, args);
    if (std.mem.eql(u8, subcommand, "drop")) return dropTable(client, args);
    if (std.mem.eql(u8, subcommand, "list")) return listTablesWithArgs(allocator, io, client, args);
    if (std.mem.eql(u8, subcommand, "get")) return getTable(allocator, io, client, args);

    if (std.mem.startsWith(u8, subcommand, "--")) {
        return runWithFlags(allocator, io, client, subcommand, args);
    }

    cli.fatal("unknown table subcommand: {s}", .{subcommand});
}

fn runWithFlags(allocator: std.mem.Allocator, io: std.Io, client: *antfly_client.AntflyClient, first_arg: []const u8, args: *std.process.Args.Iterator) !void {
    var table_name: ?[]const u8 = null;
    var output: ?[]const u8 = null;
    var current_arg: ?[]const u8 = first_arg;

    while (current_arg) |arg| : (current_arg = args.next()) {
        if (std.mem.eql(u8, arg, "--table") or std.mem.eql(u8, arg, "-t")) {
            if (table_name != null) cli.fatal("--table may only be provided once", .{});
            table_name = args.next() orelse cli.fatal("{s} requires a value", .{arg});
        } else if (std.mem.eql(u8, arg, "--output") or std.mem.eql(u8, arg, "-o")) {
            if (output != null) cli.fatal("--output may only be provided once", .{});
            output = args.next() orelse cli.fatal("{s} requires a value", .{arg});
        } else {
            cli.fatal("unknown table option: {s}", .{arg});
        }
    }

    if (output) |value| {
        if (!std.mem.eql(u8, value, "json")) {
            cli.fatal("only JSON output is supported for table", .{});
        }
    }

    if (table_name) |name| {
        return getTableByName(allocator, io, client, name);
    }
    return listTablesMode(allocator, io, client, if (output != null) .json else .summary);
}

fn createTable(allocator: std.mem.Allocator, io: std.Io, client: *antfly_client.AntflyClient, args: *std.process.Args.Iterator) !void {
    var table_name: ?[]const u8 = null;
    var shards: ?i64 = null;
    var file_path: ?[]const u8 = null;
    var schema_json: ?[]const u8 = null;
    var inline_indexes = std.ArrayListUnmanaged([]const u8).empty;
    defer inline_indexes.deinit(allocator);

    while (args.next()) |arg| {
        const option = tableCreateOption(arg) orelse cli.fatal("unknown table create option: {s}", .{arg});
        switch (option) {
            .table => {
                if (table_name != null) cli.fatal("--table may only be provided once", .{});
                table_name = args.next() orelse cli.fatal("{s} requires a value", .{arg});
            },
            .shards => {
                if (shards != null) cli.fatal("--shards may only be provided once", .{});
                const raw = args.next() orelse cli.fatal("--shards requires a value", .{});
                shards = std.fmt.parseInt(i64, raw, 10) catch cli.fatal("invalid --shards value: {s}", .{raw});
                if (shards.? <= 0) cli.fatal("--shards must be greater than zero", .{});
            },
            .file => {
                if (file_path != null) cli.fatal("--file may only be provided once", .{});
                file_path = args.next() orelse cli.fatal("{s} requires a value", .{arg});
            },
            .index => {
                const raw = args.next() orelse cli.fatal("--index requires a JSON value", .{});
                inline_indexes.append(allocator, raw) catch |err| cli.fatal("recording --index: {}", .{err});
            },
            .schema => {
                if (schema_json != null) cli.fatal("--schema may only be provided once", .{});
                schema_json = args.next() orelse cli.fatal("--schema requires a JSON value", .{});
            },
        }
    }

    const name = table_name orelse cli.fatal("--table is required", .{});

    var body = antfly_client.types.CreateTableRequest{};
    if (file_path) |path| {
        if (inline_indexes.items.len != 0) {
            cli.fatal("--index with --file is not supported; put indexes in {s}", .{path});
        }
        if (schema_json != null) {
            cli.fatal("--schema with --file is not supported; put schema in {s}", .{path});
        }
        const file_data = cli.readFileAlloc(io, allocator, path, 10 * 1024 * 1024) catch |err| {
            cli.fatal("reading config file {s}: {}", .{ path, err });
        };
        defer allocator.free(file_data);
        var parsed = std.json.parseFromSlice(antfly_client.types.CreateTableRequest, allocator, file_data, .{}) catch |err| {
            cli.fatal("parsing config file {s}: {}", .{ path, err });
        };
        defer parsed.deinit();
        if (shards != null) {
            cli.fatal("--shards with --file is not supported; put num_shards in {s}", .{path});
        }
        try client.createTable(name, parsed.value);
        std.debug.print("Create table command successful.\n", .{});
        return;
    }

    if (shards) |s| body.num_shards = s;
    var arena = std.heap.ArenaAllocator.init(allocator);
    defer arena.deinit();
    if (schema_json) |raw| {
        body.schema = parseInlineSchemaConfig(arena.allocator(), raw) catch |err| cli.fatal("invalid JSON for --schema: {}", .{err});
    }
    if (inline_indexes.items.len != 0) {
        var indexes = std.json.ArrayHashMap(antfly_client.types.CreateIndexRequest){};
        for (inline_indexes.items) |raw| {
            const index = parseInlineIndexConfig(arena.allocator(), raw) catch |err| cli.fatal("invalid JSON for --index: {}", .{err});
            putInlineIndexConfig(arena.allocator(), &indexes, index) catch |err| switch (err) {
                error.EmptyIndexName => cli.fatal("--index name must not be empty", .{}),
                error.DuplicateIndexName => cli.fatal("duplicate --index definition: {s}", .{index.name}),
                else => cli.fatal("recording --index {s}: {}", .{ index.name, err }),
            };
        }
        body.indexes = indexes;
    }

    try client.createTable(name, body);
    std.debug.print("Create table command successful.\n", .{});
}

fn dropTable(client: *antfly_client.AntflyClient, args: *std.process.Args.Iterator) !void {
    var table_name: ?[]const u8 = null;
    while (args.next()) |arg| {
        if (std.mem.eql(u8, arg, "--table") or std.mem.eql(u8, arg, "-t")) {
            if (table_name != null) cli.fatal("--table may only be provided once", .{});
            table_name = args.next() orelse cli.fatal("{s} requires a value", .{arg});
        } else {
            cli.fatal("unknown table drop option: {s}", .{arg});
        }
    }
    const name = table_name orelse cli.fatal("--table is required", .{});
    try client.dropTable(name);
    std.debug.print("Drop table command successful.\n", .{});
}

fn listTables(allocator: std.mem.Allocator, io: std.Io, client: *antfly_client.AntflyClient) !void {
    return listTablesMode(allocator, io, client, .summary);
}

const ListOutput = enum { summary, json };

fn listTablesWithArgs(allocator: std.mem.Allocator, io: std.Io, client: *antfly_client.AntflyClient, args: *std.process.Args.Iterator) !void {
    var output: ListOutput = .summary;
    var explicitly_selected = false;
    while (args.next()) |arg| {
        if (std.mem.eql(u8, arg, "--verbose")) {
            if (explicitly_selected) cli.fatal("use only one of --verbose or --output json", .{});
            output = .json;
            explicitly_selected = true;
        } else if (std.mem.eql(u8, arg, "--output") or std.mem.eql(u8, arg, "-o")) {
            if (explicitly_selected) cli.fatal("use only one of --verbose or --output json", .{});
            const value = args.next() orelse cli.fatal("--output requires json", .{});
            if (!std.mem.eql(u8, value, "json")) cli.fatal("only JSON output is supported for table", .{});
            output = .json;
            explicitly_selected = true;
        } else {
            cli.fatal("unknown table list option: {s}", .{arg});
        }
    }
    return listTablesMode(allocator, io, client, output);
}

fn listTablesMode(allocator: std.mem.Allocator, io: std.Io, client: *antfly_client.AntflyClient, output: ListOutput) !void {
    var resp = try client.listTables();
    defer resp.deinit();
    cli.expectHttpSuccess(resp);
    if (resp.data) |parsed| {
        if (output == .json) return cli.writeJson(allocator, io, parsed.value);
        cli.writeStdout(io, "NAME\tSHARDS\tINDEXES\tSTORAGE\n");
        for (parsed.value) |table| {
            const storage = if (table.storage_status.empty orelse false)
                "empty"
            else if (table.storage_status.disk_usage != null)
                "used"
            else
                "unknown";
            const line = try std.fmt.allocPrint(
                allocator,
                "{s}\t{d}\t{d}\t{s}\n",
                .{ table.name, table.shards.map.count(), table.indexes.map.count(), storage },
            );
            defer allocator.free(line);
            cli.writeStdout(io, line);
        }
    }
}

fn getTable(allocator: std.mem.Allocator, io: std.Io, client: *antfly_client.AntflyClient, args: *std.process.Args.Iterator) !void {
    var table_name: ?[]const u8 = null;
    while (args.next()) |arg| {
        if (std.mem.eql(u8, arg, "--table") or std.mem.eql(u8, arg, "-t")) {
            if (table_name != null) cli.fatal("--table may only be provided once", .{});
            table_name = args.next() orelse cli.fatal("{s} requires a value", .{arg});
        } else {
            cli.fatal("unknown table get option: {s}", .{arg});
        }
    }
    const name = table_name orelse cli.fatal("--table is required", .{});
    return getTableByName(allocator, io, client, name);
}

fn getTableByName(allocator: std.mem.Allocator, io: std.Io, client: *antfly_client.AntflyClient, name: []const u8) !void {
    var resp = try client.getTable(name);
    defer resp.deinit();
    cli.expectHttpSuccess(resp);
    if (resp.data) |parsed| {
        try cli.writeJson(allocator, io, parsed.value);
    }
}

test "table create option classification fails closed" {
    try std.testing.expectEqual(TableCreateOption.index, tableCreateOption("--index").?);
    try std.testing.expectEqual(TableCreateOption.schema, tableCreateOption("--schema").?);
    try std.testing.expect(tableCreateOption("--indx") == null);
}

test "table create inline schema and repeatable indexes build a typed request" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    const schema = try parseInlineSchemaConfig(alloc,
        \\{"default_type":"doc","enforce_types":false}
    );
    try std.testing.expectEqualStrings("doc", schema.default_type.?);

    var indexes = std.json.ArrayHashMap(antfly_client.types.CreateIndexRequest){};
    const dense = try parseInlineIndexConfig(alloc,
        \\{"name":"title_body","type":"embeddings","field":"body","dimension":512}
    );
    try putInlineIndexConfig(alloc, &indexes, dense);
    const text = try parseInlineIndexConfig(alloc,
        \\{"name":"body_text","type":"full_text"}
    );
    try putInlineIndexConfig(alloc, &indexes, text);

    const request = antfly_client.types.CreateTableRequest{
        .num_shards = 3,
        .schema = schema,
        .indexes = indexes,
    };
    const encoded = try std.json.Stringify.valueAlloc(std.testing.allocator, request, .{});
    defer std.testing.allocator.free(encoded);
    try ant_json.testing.expectSubsetJsonText(
        std.testing.allocator,
        "{\"num_shards\":3,\"indexes\":{\"title_body\":{\"type\":\"embeddings\",\"dimension\":512},\"body_text\":{\"type\":\"full_text\"}}}",
        encoded,
    );
}

test "table create inline configs reject malformed unknown and duplicate definitions" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    try std.testing.expectError(error.UnexpectedEndOfInput, parseInlineIndexConfig(alloc, "{"));
    try std.testing.expectError(error.UnknownField, parseInlineIndexConfig(alloc,
        \\{"name":"bad","type":"embeddings","typo":true}
    ));

    var indexes = std.json.ArrayHashMap(antfly_client.types.CreateIndexRequest){};
    const first = try parseInlineIndexConfig(alloc,
        \\{"name":"duplicate","type":"full_text"}
    );
    try putInlineIndexConfig(alloc, &indexes, first);
    const second = try parseInlineIndexConfig(alloc,
        \\{"name":"duplicate","type":"embeddings","dimension":3,"external":true}
    );
    try std.testing.expectError(error.DuplicateIndexName, putInlineIndexConfig(alloc, &indexes, second));

    try std.testing.expectError(error.UnknownField, parseInlineSchemaConfig(alloc,
        \\{"unknown_schema_field":true}
    ));
}

test "table create sends the exact quickstart inline index through the HTTP client" {
    const alloc = std.testing.allocator;
    const io = std.testing.io;

    const Assert = struct {
        fn request(info: httpx.testing_mod.RequestInfo) !void {
            try ant_json.testing.expectSubsetJsonText(
                std.testing.allocator,
                "{\"indexes\":{\"title_body\":{\"type\":\"embeddings\",\"template\":\"{{title}} {{body}}\",\"embedder\":{\"model\":\"antflydb/clipclap\"},\"chunker\":{\"text\":{\"target_tokens\":200}}}}}",
                info.body,
            );
            var parsed = try ant_json.parseFromSlice(ant_json.Value, std.testing.allocator, info.body, .{});
            defer parsed.deinit();
            const root = parsed.value.object;
            const indexes = root.get("indexes") orelse return error.MissingIndexes;
            const title_body = indexes.object.get("title_body") orelse return error.MissingTitleBody;
            // The map key owns the index name in the breaking-major table
            // contract. Do not regress to two independently mutable names.
            try std.testing.expect(title_body.object.get("name") == null);
        }
    };

    var server = try httpx.TestServer.start(alloc, io, &.{.{
        .method = .POST,
        .path = "/db/v1/tables/wikipedia",
        .respond = .{ .body = "{\"name\":\"wikipedia\",\"indexes\":{},\"shards\":{}}" },
        .assert_request = Assert.request,
    }});
    defer server.deinit();

    var http = httpx.Client.initWithConfig(alloc, io, .{ .keep_alive = false });
    defer http.deinit();
    var client = try antfly_client.AntflyClient.init(alloc, &http, server.baseUrl());
    defer client.deinit();
    var succeeded = std.atomic.Value(bool).init(false);
    var group = std.Io.Group.init;
    const ClientTask = struct {
        fn run(test_io: std.Io, c: *antfly_client.AntflyClient, success: *std.atomic.Value(bool)) std.Io.Cancelable!void {
            var argv = [_][*:0]const u8{
                "--table",
                "wikipedia",
                "--index",
                \\{"name":"title_body","type":"embeddings","template":"{{title}} {{body}}","embedder":{"provider":"antfly","model":"antflydb/clipclap"},"chunker":{"provider":"antfly","text":{"target_tokens":200,"overlap_tokens":25}}}
                ,
            };
            var args = std.process.Args.Iterator.init(.{ .vector = argv[0..] });
            createTable(std.testing.allocator, test_io, c, &args) catch return;
            success.store(true, .release);
        }
    };
    try group.concurrent(io, ClientTask.run, .{ io, &client, &succeeded });
    try server.handleOne();
    try group.await(io);
    try std.testing.expect(succeeded.load(.acquire));
}
